### ISSUE TITLE
Functionality - Installation

### ISSUE THREAD

**Issue by andrewfowlie** [ejhigson/dyPolyChord#3]
Hi,

For my [review](https://github.com/openjournals/joss-reviews/issues/965), I am trying to check-off this one,

> ### Functionality
> - [ ] **Installation:** Does installation proceed as outlined in the documentation?

I have two problems. 

1. It is very easy to install your package, run the tests, and still miss the fact that PolyChord and/or PyPolyChord are completely missing. I think this could be improved. E.g., since this is a Python package and PolyChord v1.14 is required, why not make PyPolyChord a requirement and indicate that it is missing?

2. Without PyPolyChord, I could run the tests quite quickly (and a few of them were skipped). Upon installing PyPolyChord, I ran the tests. I saw lots of `Polychord Warning: Non deterministic loglikelihood`, and it ran for a long time without finishing. A note about this warning and some indication about how long this step might take would be helpful.

**Comment by ejhigson** [comment]
> 1. It is very easy to install your package, run the tests, and still miss the fact that PolyChord and/or PyPolyChord are completely missing. I think this could be improved. E.g., since this is a Python package and PolyChord v1.14 is required, why not make PyPolyChord a requirement and indicate that it is missing?

The tests are deliberately designed to run without PyPolyChord so people who only use C++ and Fortran likelihoods can perform tests without installing PyPolyChord. Similarly the tests check the code required for C++ and Fortran likelihoods without needing a compiled executable (which would be a pain for the user). Another reason for this approach is that it allows the use of continuous integration with travis CI, which would otherwise be very difficult or impossible due to the need to download PolyChord from sourceforge with a login.

You are correct that this is a bit confusing - I have added a warning to the tests which triggers if PyPolyChord is not installed, reminding the user that it isnt installed but is only needed if they want to use Python likelihoods.

**Comment by ejhigson** [comment]
> 2\. Without PyPolyChord, I could run the tests quite quickly (and a few of them were skipped). Upon installing PyPolyChord, I ran the tests. I saw lots of `Polychord Warning: Non deterministic loglikelihood`, and it ran for a long time without finishing. A note about this warning and some indication about how long this step might take would be helpful.

This is not intended to happen! What version of PolyChord and Python are you using? And do you know which test this occurs in?

Do you get any warnings saying something like "This indicates that your PolyChord install\'s random number  generator is different to the one used to calculate the expected results, so I will skip tests of exact numerical values. If all the other tests pass then everything should work ok.'?

The tests use very low nlive and num_repeat settings to make them fast, which can cause this problem. However they also use random seeding which should prevent it from occurring. I suspect the seeding dosn't work on your machine and hence the tests need to be changed.

**Comment by andrewfowlie** [comment]
Thanks for your responses. I’m happy that the first one is now addressed. About the second one, I’ll have to get back to you tomorrow after running the tests again. I’ll try to isolate which test caused that behaviour, whether other messages were printed, and tell you my version numbers etc.

**Comment by ejhigson** [comment]
sounds great, thank you very much!

**Comment by andrewfowlie** [comment]
Sorry, I spoke too soon. I just checked the change https://github.com/ejhigson/dyPolyChord/commit/6ba7ec0a02e645ff7b3e17890e8785f2b24de417. It looks good, except that the warning could be printed if PyPolyChord is present but mpi4py isn’t.

**Comment by ejhigson** [comment]
Well spotted! Fix in fc4a90ce51c4fd1affab284671b9393f97401081

**Comment by andrewfowlie** [comment]
# Problem

After git pulling today (latest commit fc4a90ce51c4fd1affab284671b9393f97401081), I found the same as yesterday,
```
andrew@andrew-laptop:~/repos/dyPolyChord$ nosetests &> nosetests.log
andrew@andrew-laptop:~/repos/dyPolyChord$ more nosetests.log 
............................F/home/andrew/repos/dyPolyChord/dyPolyChord/nlive_allocation.py:49: UserWarning: Smoothing has added turning points to nlive allocation when dynamic_goal=0. I am using the unsmoothed nlives instead.
  'instead.'), UserWarning)
Polychord Warning: Non deterministic loglikelihood
Polychord Warning: Non deterministic loglikelihood
Polychord Warning: Non deterministic loglikelihood
```
The warning is repeated many times and it doesn't finish in a reasonable time. 

# Version numbers etc

My versions are
```
python -V
Python 2.7.15rc1

>>> numpy.__version__
'1.15.1'
>>> nestcheck.__version__
'0.1.8'
>>> scipy.__version__
'1.1.0'

PolyChord_v1.15.tar.gz
```
The installations all went OK, with the exception of PyPolyChord, where I had to `export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libmpi_cxx.so` to avoid a missing symbol problem with MPI. I believe though that my installation is fine, though unfortunately there aren't any tests, e.g.,
```
andrew@andrew-laptop:~/codes/PolyChord$ python run_PyPolyChord.py 
...
 ____________________________________________________ 
|                                                    |
| ndead  =         1039                              |
| log(Z) =           -1.73781 +/-            0.24593 |
|____________________________________________________|
Install matplotlib and getdist for plotting examples

``` 

# Isolating tests that fail

By investigating further I found that the test that doesn't finish in a reasonable time is `test_dynamic_evidence`. When I remove that one by adding `return True`
```
andrew@andrew-laptop:~/repos/dyPolyChord$ git diff
diff --git a/tests/tests.py b/tests/tests.py
index 8067f7f..4194899 100644
--- a/tests/tests.py
+++ b/tests/tests.py
@@ -128,6 +128,7 @@ class TestRunDyPolyChordNumers(unittest.TestCase):
 
     def test_dynamic_evidence(self):
         """Test numerical results for nested sampling with dynamic_goal=0."""
+        return True
         dynamic_goal = 0
         dyPolyChord.run_dypolychord(
             self.run_func, dynamic_goal, self.settings,
```
I find though that two other tests fail
```
andrew@andrew-laptop:~/repos/dyPolyChord$ nosetests 
............................F.F.....
======================================================================
FAIL: Test numerical results for nested sampling with
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 163, in test_dynamic_both_evidence_and_param
    self.assertEqual(e.count_samples(run), 556)
AssertionError: 426 != 556

======================================================================
FAIL: Test numerical results for nested sampling with dynamic_goal=1.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 182, in test_dynamic_param
    places=12)
AssertionError: 0.06839158405509836 != 0.029776073688009493 within 12 places

----------------------------------------------------------------------
Ran 36 tests in 4.719s

FAILED (failures=2)
```

# Tweaking `precision_criterion`

i reduced the `precision_criterion` as your code indicates that this could be related to the non-deterministic errors.

```
andrew@andrew-laptop:~/repos/dyPolyChord$ git diff
diff --git a/tests/tests.py b/tests/tests.py
index 8067f7f..c7d3513 100644
--- a/tests/tests.py
+++ b/tests/tests.py
@@ -111,7 +111,7 @@ class TestRunDyPolyChordNumers(unittest.TestCase):
             # errors. These occur due in the low dimension and low and nlive
             # cases we use for fast testing as runs sometimes get very close
             # to the peak where the likelihood becomes approximately constant.
-            'precision_criterion': 0.01,
+            'precision_criterion': 0.001,
             'seed': 1,
             'max_ndead': -1,
             'base_dir': TEST_CACHE_DIR,
```
I find that it then runs alright, but still fails
```
andrew@andrew-laptop:~/repos/dyPolyChord$ nosetests
............................F/home/andrew/repos/dyPolyChord/dyPolyChord/nlive_allocation.py:49: UserWarning: Smoothing has added turning points to nlive allocation when dynamic_goal=0. I am using the unsmoothed nlives instead.
  'instead.'), UserWarning)
FF.....
======================================================================
FAIL: Test numerical results for nested sampling with
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 162, in test_dynamic_both_evidence_and_param
    self.assertEqual(e.count_samples(run), 556)
AssertionError: 525 != 556

======================================================================
FAIL: Test numerical results for nested sampling with dynamic_goal=0.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 143, in test_dynamic_evidence
    self.assertEqual(e.count_samples(run), 548)
AssertionError: 554 != 548

======================================================================
FAIL: Test numerical results for nested sampling with dynamic_goal=1.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 181, in test_dynamic_param
    places=12)
AssertionError: 0.09796131954595269 != 0.029776073688009493 within 12 places

----------------------------------------------------------------------
Ran 36 tests in 5.565s

FAILED (failures=3)
```

**Comment by ejhigson** [comment]
Thank you very much for this!!

I think the problem is due to the previous settings not being very robust when the random seeing does not give them exactly the result expected. I have changed the settings in a manner which should solve this problem in 70a9e62f5033665b4c2ef936c7643fd7c8fad517, and updated the values the tests expect to the ones I get on my machine (the tests checking the result values caused the failures you show above). The tests are a little slower but they still don't take long so I think this is fine.

Let me know if this works for you!

**Comment by andrewfowlie** [comment]
```
andrew@andrew-laptop:~/repos/dyPolyChord$ git pull
Updating fc4a90c..70a9e62
Fast-forward
 tests/tests.py | 34 ++++++++++++++++++----------------
 1 file changed, 18 insertions(+), 16 deletions(-)
andrew@andrew-laptop:~/repos/dyPolyChord$ export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libmpi_cxx.so
andrew@andrew-laptop:~/repos/dyPolyChord$ nosetests
............................FFF.....
======================================================================
FAIL: Test numerical results for nested sampling with
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 164, in test_dynamic_both_evidence_and_param
    self.assertEqual(e.count_samples(run), 1093)
AssertionError: 920 != 1093

======================================================================
FAIL: Test numerical results for nested sampling with dynamic_goal=0.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 145, in test_dynamic_evidence
    self.assertEqual(e.count_samples(run), 1169)
AssertionError: 924 != 1169

======================================================================
FAIL: Test numerical results for nested sampling with dynamic_goal=1.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrew/repos/dyPolyChord/tests/tests.py", line 183, in test_dynamic_param
    places=12)
AssertionError: -0.029540931478677097 != -0.05323120028149568 within 12 places

----------------------------------------------------------------------
Ran 36 tests in 33.492s

FAILED (failures=3)

```

**Comment by ejhigson** [comment]
Gotcha! The random seeding is working ok - the problem was a rogue division of integers when computing how many samples to use

`samp_tot = init_run['logl'].shape[0] * (nlive_const / ninit)`

With a fresh download of the latest PolyChord, running in python 2, the fix in 84ee43ac91e2ddd3f88bd92847de87d202d8398d is now working for me whereas without the fix I got the exact same errors as you. I work almost exclusively in python 3 and this must have somehow slipped past my previous python2 testing.

Thank you very much for your help finding this @andrewfowlie! If you confirm this is working for you I will close the issue

**Comment by andrewfowlie** [comment]
```
andrew@andrew-laptop:~/repos/dyPolyChord$ nosetests
....................................
----------------------------------------------------------------------
Ran 36 tests in 28.845s

OK
```
Great! Thanks for investigating this.
